File: shelfy-frontend/src/lib/telegram-html.ts
/** Subset HTML compatible con Telegram parse_mode=HTML */

const ALLOWED = new Set(["b", "i", "u", "s", "code", "pre"]);

/** Secuencias literales \\n en DB (seed SQL sin E'…') → saltos reales. */
function unescapeLiteralBackslashSequences(text: string): string {
  return text.replace(/\\([nrt])/g, (_, ch: string) => {
    if (ch === "n") return "\n";
    if (ch === "r") return "\r";
    return "\t";
  });
}

/** Limpia HTML guardado (DB) antes de editar o previsualizar. */
export function sanitizeStoredTelegramHtml(text: string): string {
  if (!text) return "";
  return unescapeLiteralBackslashSequences(text)
    .replace(/\r\n/g, "\n")
    .replace(/\r/g, "\n")
    .replace(/<br\s*\/?>/gi, "\n")
    .replace(/<\/div>/gi, "\n")
    .replace(/<div[^>]*>/gi, "")
    .replace(/<\/p>/gi, "\n")
    .replace(/<p[^>]*>/gi, "")
    .replace(/<(b|i|u|s|code|pre)\b[^>]*>/gi, "<$1>")
    .replace(/<\/(b|i|u|s|code|pre)\b[^>]*>/gi, "</$1>")
    .replace(/<\/?(?:span|font)\b[^>]*>/gi, "")
    .replace(/&nbsp;/gi, " ")
    .replace(/\u200b/g, "")
    .replace(/\n{3,}/g, "\n\n")
    .replace(/\n (?=[^\s•])/g, "\n");
}

/** Normaliza HTML del contentEditable → string Telegram (saltos = \\n). */
export function normalizeTelegramHtml(raw: string): string {
  if (!raw) return "";

  let html = unescapeLiteralBackslashSequences(raw)
    .replace(/\r\n/g, "\n")
    .replace(/\r/g, "\n")
    .replace(/<strong\b[^>]*>/gi, "<b>")
    .replace(/<\/strong>/gi, "</b>")
    .replace(/<em\b[^>]*>/gi, "<i>")
    .replace(/<\/em>/gi, "</i>")
    .replace(/<br\s*\/?>/gi, "\n")
    .replace(/<\/div>/gi, "\n")
    .replace(/<div[^>]*>/gi, "")
    .replace(/<\/p>/gi, "\n")
    .replace(/<p[^>]*>/gi, "")
    .replace(/<\/?(?:span|font)\b[^>]*>/gi, "")
    .replace(/&nbsp;/gi, " ")
    .replace(/\u200b/g, "");

  html = html.replace(/<(b|i|u|s|code|pre)\b[^>]*>/gi, "<$1>");
  html = html.replace(/<\/(b|i|u|s|code|pre)\b[^>]*>/gi, "</$1>");

  // Quitar tags no permitidos (conservar contenido)
  html = html.replace(/<(\/?)([\w]+)([^>]*)>/gi, (match, slash, tag) => {
    const t = tag.toLowerCase();
    if (ALLOWED.has(t)) return match;
    return "";
  });

  html = html.replace(/\n{3,}/g, "\n\n");
  html = html.replace(/\n (?=[^\s•])/g, "\n");

  // Colapsar tags anidados del mismo tipo consecutivos vacíos
  html = html.replace(/(<(\w+)>)\s*(<\2>)+/gi, "$1");
  html = html.replace(/(<\/(\w+)>)\s*(<\/\2>)+/gi, "$1");

  // No trim(): preserva \n inicial/final en plantillas dinámicas (filas ranking, items objetivos).
  return html.replace(/^\s+/, "").replace(/[ \t]+$/, "");
}

/** Telegram HTML → HTML seguro para render en preview (contentEditable). */
export function telegramHtmlToRenderHtml(text: string): string {
  if (!text) return "";
  const clean = sanitizeStoredTelegramHtml(text);
  const escaped = clean
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");

  return escaped
    .replace(/&lt;b&gt;/gi, "<b>")
    .replace(/&lt;\/b&gt;/gi, "</b>")
    .replace(/&lt;i&gt;/gi, "<i>")
    .replace(/&lt;\/i&gt;/gi, "</i>")
    .replace(/&lt;u&gt;/gi, "<u>")
    .replace(/&lt;\/u&gt;/gi, "</u>")
    .replace(/&lt;s&gt;/gi, "<s>")
    .replace(/&lt;\/s&gt;/gi, "</s>")
    .replace(/&lt;code&gt;/gi, "<code>")
    .replace(/&lt;\/code&gt;/gi, "</code>")
    .replace(/\n/g, "<br>");
}

export function isTelegramHtmlEmpty(text: string): boolean {
  const plain = text.replace(/<[^>]+>/g, "").replace(/&nbsp;/g, " ").trim();
  return plain.length === 0;
}

File: CenterMind/core/telegram_html.py
# -*- coding: utf-8 -*-
"""
Normalización HTML Telegram (parse_mode=HTML).
Telegram usa \\n para saltos de línea — no <br>.
"""
from __future__ import annotations

import re

_ALLOWED_TAGS = frozenset({"b", "i", "u", "s", "code", "pre"})
_TAG_ATTR_RE = re.compile(r"<(b|i|u|s|code|pre)\b[^>]*>", re.I)
_TAG_CLOSE_RE = re.compile(r"</(b|i|u|s|code|pre)\b[^>]*>", re.I)
_DISALLOWED_TAG_RE = re.compile(r"</?(?:span|font|div|p)\b[^>]*>", re.I)
_BR_RE = re.compile(r"<br\s*/?>", re.I)
# Secuencias literales \n (dos chars) por seed SQL sin E'…' o doble-escape JSON
_LITERAL_ESCAPES_RE = re.compile(r"\\([nrt])")


def _unescape_literal_backslash_sequences(text: str) -> str:
    """Convierte \\n, \\r, \\t literales → saltos/tabs reales."""

    def repl(m: re.Match[str]) -> str:
        ch = m.group(1)
        if ch == "n":
            return "\n"
        if ch == "r":
            return "\r"
        return "\t"

    return _LITERAL_ESCAPES_RE.sub(repl, text)


def repair_telegram_message_html(text: str) -> str:
    """
    Convierte HTML de editor WYSIWYG → HTML válido para Telegram.
    - <br>, </div>, </p> → \\n
    - Elimina ZWSP y tags no permitidos (conserva contenido).
    - Normaliza tags permitidos sin atributos.
    """
    if not text:
        return ""

    s = _unescape_literal_backslash_sequences(text)
    s = s.replace("\r\n", "\n").replace("\r", "\n")
    s = _BR_RE.sub("\n", s)
    s = re.sub(r"</div>", "\n", s, flags=re.I)
    s = re.sub(r"<div[^>]*>", "", s, flags=re.I)
    s = re.sub(r"</p>", "\n", s, flags=re.I)
    s = re.sub(r"<p[^>]*>", "", s, flags=re.I)
    s = _DISALLOWED_TAG_RE.sub("", s)
    s = s.replace("\u200b", "").replace("&nbsp;", " ")
    s = _TAG_ATTR_RE.sub(lambda m: f"<{m.group(1).lower()}>", s)
    s = _TAG_CLOSE_RE.sub(lambda m: f"</{m.group(1).lower()}>", s)

    # Quitar cualquier otro tag HTML (conservar texto)
    s = re.sub(r"<(?!/?(?:b|i|u|s|code|pre)\b)[^>]+>", "", s, flags=re.I)

    s = re.sub(r"\n{3,}", "\n\n", s)
    # Editor manual: a veces deja un espacio suelto tras \\n (no indentación con •).
    s = re.sub(r"\n (?=[^\s•])", "\n", s)
    # Solo espacios/tabs en bordes — no strip() ni lstrip() genérico (borran \\n de plantillas).
    return s.lstrip(" \t").rstrip(" \t")


def message_needs_linebreak_repair(text: str) -> bool:
    """True si el texto guardado tiene artefactos típicos del bug de saltos."""
    if not text:
        return False
    lower = text.lower()
    return (
        "<br" in lower
        or "\u200b" in text
        or "</div>" in lower
        or "<div" in lower
        or "</p>" in lower
        or "<p" in lower
        or _LITERAL_ESCAPES_RE.search(text) is not None
    )
